tests/store/tracking/test_sqlalchemy_store.py

```diff
@@ -255,6 +255,56 @@ def test_delete_experiment(self):
 
         self.assertEqual(len(self.store.list_experiments()), len(all_experiments) - 1)
 
+    def test_delete_experiment_with_runs(self):
+        # Create run with experiment
+        run = self._run_factory()
+        all_experiments = self.store.list_experiments()
+        len_of_experiments = 2  # default + current experiment, hence 2 experiments
+        self.assertEqual(len(all_experiments), len_of_experiments)
+
+        exp_id = run.info.experiment_id  # Find experiment id
+        run_id = run.info.run_id  # Find run id
+
+        self.store.delete_experiment(exp_id)
+
+        updated_exp = self.store.get_experiment(exp_id)
+        self.assertEqual(updated_exp.lifecycle_stage, entities.LifecycleStage.DELETED)
+
+        updated_run = self.store.get_run(run_id)
+        self.assertEqual(updated_run.info.lifecycle_stage, entities.LifecycleStage.DELETED)
+
+        self.assertEqual(len(self.store.list_experiments()), len_of_experiments - 1)
+
+    def test_restore_experiment_with_runs(self):
+        # Create run with experiment
+        run = self._run_factory()
+        all_experiments = self.store.list_experiments()
+        len_of_experiments = 2  # default + current experiment, hence 2 experiments
+        self.assertEqual(len(all_experiments), len_of_experiments)
+
+        exp_id = run.info.experiment_id  # Find experiment id
+        run_id = run.info.run_id  # Find run id
+
+        self.store.delete_experiment(exp_id)
+
+        updated_exp = self.store.get_experiment(exp_id)
+        self.assertEqual(updated_exp.lifecycle_stage, entities.LifecycleStage.DELETED)
+
+        updated_run = self.store.get_run(run_id)
+        self.assertEqual(updated_run.info.lifecycle_stage, entities.LifecycleStage.DELETED)
+
+        self.assertEqual(len(self.store.list_experiments()), len_of_experiments - 1)
+
+        self.store.restore_experiment(exp_id)
+
+        updated_exp = self.store.get_experiment(exp_id)
+        self.assertEqual(updated_exp.lifecycle_stage, entities.LifecycleStage.ACTIVE)
+
+        updated_run = self.store.get_run(run_id)
+        self.assertEqual(updated_run.info.lifecycle_stage, entities.LifecycleStage.ACTIVE)
+
+        self.assertEqual(len(all_experiments), len_of_experiments)
```

Comment:
could each run_id be validated before / after the delete and restore by doing a set `.issubset()` validation to be on the safe side instead of comparing the collection length?

Reply:
@BenWilson2 : run_id is a string, which is returned by the `self._run_factory()` method. String doesn’t have a `.issubset()` method. I can check with `in`, with `run_id` as part of `run.info.run_id`. Please do let me know if this is ok, or do suggest some other mechanism.

Reply:
That wasn't precisely what I was referring to. I'm sorry that I wasn't more clear.
The logic that I was going for was:
- create experiment and run in experiment
- get a set of run_id's present in experiment
- mark experiment for deletion
- validate all runs (should be just one) are marked for deletion
- restore experiment
- list runs in the experiment that was restored
- ensure set(run_ids of original experiment pre deletion) is a subset of set(run_ids of restored experiment)

This check just adds visibility to ensure that we're not in the future going to introduce logic that might regenerate a run_id for a restored experiment (just being cautious here).

Reply:
@BenWilson2 : I have made the changes you suggested. Kindly have a look.

Reply:
looks great! Thank you!